### PR TITLE
tests, vmi networking, make test 1780 dual stack compatible

### PR DESCRIPTION
**Pull request by oshoval**
In case the cluster is dual stack, 2 ips might exists.
Test each one of the existing ips on networking test 1780.

Use PingFromVMConsole new mechanism
instead of trace route, for simplification.

Depends on
- [x] #3957 
- [x] #3958
- [x] #4000  
- [x] #4005 

Signed-off-by: Or Shoval <oshoval@redhat.com>

```release-note
None
```

### CODE REVIEW

**Comment by oshoval** [comment]
/cc @AlonaKaplan 
preliminary change, waiting for the k8s-1.18 , 1.17 to pass 

Reason i didnt use describeTable, is that the straight forward way is to let the test first create the VM
(which is happening "after" the table is taking place"
and then test each avialable field, according its type

else we would have 3 lines in the table we already have X 2 per each possible ip,
or other methods i can think on which arent good imo,
unless you have a better method, i believe its natural and good method, non intrusive, generic.

**Comment by oshoval** [comment]
known that something wrong here, checking because it might be important for the scheme PR

File: tests/vmi_networking_test.go
Comment (oshoval) [inline tests/vmi_networking_test.go:753]: after refactor of the vm PR we wont have this if,
but the skip will be under if ipv6 condition, since this way it it skip exactly what needed and be explicit that it doesnt skip IPv4.
the skip it just until the core changes of the dual stack are fixed anyway.

**Comment by oshoval** [comment]
Rebased,
cherry-pick the new version of VM scheme,
updated the skip as it should be

File: tests/vmi_networking_test.go
Comment (AlonaKaplan) [inline tests/vmi_networking_test.go:727]: I would remove https://github.com/kubevirt/kubevirt/pull/3957 from the skip message. All this PR is depends on it. So I would update the PR message and add a hold to the PR until the mentined PR is merged.

Comment (oshoval) [reply tests/vmi_networking_test.go:727]: hold you mean here or on #3957 ?
because i think i can remove 3957 mentioned here, but we dont need any hold?
once all the other needed PRs are merged, the skip will be safe the remove and whoever need this code, will have it already while developing, wdyt?

beside that the method is fine that we will use it on all the other tests that have the same phenomena ?

Comment (oshoval) [reply tests/vmi_networking_test.go]: updated title, left just 432

**Comment by oshoval** [comment]
/hold 
until  #3958 is merged

File: tests/vmi_networking_test.go
Comment (AlonaKaplan) [inline tests/vmi_networking_test.go:727]: https://github.com/kubevirt/kubevirtci/pull/432 was merged.

Comment (oshoval) [reply tests/vmi_networking_test.go:727]: ye but there was a kubevirt bump ?
and its the only one we need ? i wanted to make sure manually that it works before, and see what PRs need to be merged (for example vm scheme), before i remove the hold

Comment (oshoval) [reply tests/vmi_networking_test.go:727]: it doesnt work with the following
VM scheme PR
bump to have 432 in kubevirt itself
remove the skip of 432

we need https://github.com/kubevirt/kubevirt/pull/3957 at least
so we either wait for it, or add the skip according that one.
i vote for the 2nd, i dont see a reason to wait if its ok please.

File: tests/vmi_networking_test.go
Comment (oshoval) [reply tests/vmi_networking_test.go]: i dont think we should wait, see please https://github.com/kubevirt/kubevirt/pull/3989#discussion_r472213579

File: tests/vmi_networking_test.go
Comment (AlonaKaplan) [reply tests/vmi_networking_test.go:727]: We should wait until #3957, #4000 and #3958 are merged. Meanwhile, this PR can depend on those PRs.

File: tests/vmi_networking_test.go
Comment (AlonaKaplan) [inline tests/vmi_networking_test.go:725]: Please add `By(IPv4)` and `By(IPv6)`.

File: tests/vmi_networking_test.go
Comment (EdDev) [inline tests/vmi_networking_test.go:763]: This should not be in the for-loop.

File: tests/vmi_networking_test.go
Comment (oshoval) [reply tests/vmi_networking_test.go:727]: why should we wait ? usually we just add a skip until and do move forward.

File: tests/vmi_networking_test.go
Comment (oshoval) [reply tests/vmi_networking_test.go:725]: added already after it checks the type on lines
https://github.com/kubevirt/kubevirt/pull/3989/files#diff-f911024a391a99cfdef2241f409a4748R731
https://github.com/kubevirt/kubevirt/pull/3989/files#diff-f911024a391a99cfdef2241f409a4748R740

will see if more needed in the beginning of each section

File: tests/vmi_networking_test.go
Comment (EdDev) [reply tests/vmi_networking_test.go:727]: Because we do not know if it works or not, or better phrase it as: "we do not **add** dead code".

Comment (oshoval) [reply tests/vmi_networking_test.go:727]: I meant that i will check locally, be sure that it works, as i did before and stopped when i found something that changed the story. 
but nvm

File: tests/vmi_networking_test.go
Comment (oshoval) [reply tests/vmi_networking_test.go]: Thanks done, as i wrote i stopped checking when i saw preliminary needs.

File: tests/vmi_networking_test.go
Comment (oshoval) [reply tests/vmi_networking_test.go:726]: done

**Comment by oshoval** [comment]
Tested with the 4 mentioned PRs, it works,
we can either merge it with a skip, or just wait for them.

File: tests/vmi_networking_test.go
Comment (oshoval) [inline tests/vmi_networking_test.go:730]: tested locally with the new ping @EdDev created, so i can update it to ping if needed now.

File: tests/vmi_networking_test.go
Comment (alonSadan) [inline tests/vmi_networking_test.go:724]: nit: maybe using `To` instead of `Should`?
Expect(len(serverVMI.Status.Interfaces[0].IPs)).To(BeNumerically(">", 0))
better English I think

File: tests/vmi_networking_test.go
Comment (alonSadan) [inline tests/vmi_networking_test.go:762]: It should fail to connect here right?
maybe format a message here that explains what should have happen in case it fails?  
I am not sure it will be clear enough from the error returning from `CheckForTextExpecter`.
something like:
`Expect(err).ToNot(HaveOccurred(), "should fail to connect to unregistered port")`

wasn't sure about line 759 and line 755, maybe there a message will help as well, but not sure

Comment (oshoval) [reply tests/vmi_networking_test.go:762]: hmm, i just idented those line, can we take it to other PR ?
its not connected to the title and to the PR purpose

Comment (alonSadan) [reply tests/vmi_networking_test.go:762]: yes ,you are right. maybe not so related here

File: tests/vmi_networking_test.go
Comment (EdDev) [reply tests/vmi_networking_test.go:724]: Gomega docs recommends using `To` when `Expect` is used.

File: tests/vmi_networking_test.go
Comment (EdDev) [inline tests/vmi_networking_test.go:730]: Please drop the skip, if may be missed in the future reviews.

File: tests/vmi_networking_test.go
Comment (EdDev) [inline tests/vmi_networking_test.go:754]: How is this working if you call it twice? The second time it should have failed as the server is already running.

File: tests/vmi_networking_test.go
Comment (EdDev) [inline tests/vmi_networking_test.go:736]: The `By` content needs to be updated now.

**Review by EdDev — CHANGES_REQUESTED** [review]
I think this test readability is hurt now due to the for-loop.
This test needs to be refactored and adjusted to fit the complexity that the dual stack introduces.

For the sake of making sure that the dual stack really works, we can defer it to later.

File: tests/vmi_networking_test.go
Comment (oshoval) [reply tests/vmi_networking_test.go:754]: the 2nd doesnt fail (checked manually), but the command of it doesnt affect,
for example tried to change the string it output and it didnt affect anything.
also if you list the screens you see it didnt change the previous PID

I will move it just above the for, thanks

File: tests/vmi_networking_test.go
Comment (oshoval) [reply tests/vmi_networking_test.go]: Done

**Review by EdDev — COMMENTED** [review]
The changes themselves look good, I had only one nit inline.
My main concern here is that we assume there are 2 IP/s, one of which is IPv6. But we do not really know if this is the case, i.e. the tests may pass and no IPv6 was tested.

We need to add an explicit test that based on an imputed flag (or other mechanism) this expectation is checked.

File: tests/vmi_networking_test.go
Comment (oshoval) [reply tests/vmi_networking_test.go:733]: yes but Alona asked those (after the other lines were there already) so i added,
wdyt then?
its like a "title"

**Comment by oshoval** [comment]
> The changes themselves look good, I had only one nit inline.
> My main concern here is that we assume there are 2 IP/s, one of which is IPv6. But we do not really know if this is the case, i.e. the tests may pass and no IPv6 was tested.
> 
> We need to add an explicit test that based on an imputed flag (or other mechanism) this expectation is checked.

Added a suggested commit

File: tests/vmi_networking_test.go
Comment (oshoval) [inline tests/vmi_networking_test.go:903]: will lower case it

File: tests/vmi_networking_test.go
Comment (oshoval) [inline tests/vmi_networking_test.go:891]: 1.X
X can start with zero, will fix it

File: tests/vmi_networking_test.go
Comment (ormergi) [reply tests/vmi_networking_test.go]: Having all the details in one line before the tests start is much more easier to track instead of searching in the Expecter log

File: tests/vmi_networking_test.go
Comment (ormergi) [inline tests/vmi_networking_test.go:889]: Why dont you check if a cluster supports dual-stack by checking if one of kube-system pods have IPv4 and IPv6 addresses?
Eventually all providers will support dual-stack

Comment (oshoval) [reply tests/vmi_networking_test.go:889]: because if there will be a bug in the provider, the above method wont find it.
Also the check i added do support future providers, starting with 1.18
when all the lanes that are supported, support dual stack we would be able to just Expect that there are 2 IPS,
moreover, there are providers which arent k8s, like kind, for those we dont need to enforce 2, and we dont.

We should first focus on the effort to make dual stack works, imo its good enough for now.

**Comment by oshoval** [comment]
> "Having all the details in one line before the tests start is much more easier to track instead of searching in the Expecter log"

Just one ip there is constant, the rest are dynamic and will be exactly near the command that was actually failed.
I dont see a reason to add a code which is redundant and to maintain it.

File: tests/vmi_networking_test.go
Comment (oshoval) [reply tests/vmi_networking_test.go]: Just one ip there is constant, the rest are dynamic and will be exactly near the command that was actually failed.
I dont see a reason to add a code which is redundant (or not needed in case of the dynamic) and to maintain it.

File: tests/vmi_networking_test.go
Comment (ormergi) [reply tests/vmi_networking_test.go:889]: > because if there will be a bug in the provider, the above method wont find it.
> Also the check i added do support future providers, starting with 1.18
> when all the lanes that are supported, support dual stack we would be able to just Expect that there are 2 IPS,
> (there are providers which arent k8s, like kind, for those we dont need to enforce 2, and we dont)

Kubevirt tests should not take in account that there might be a bug in the provider, all tests assume the environment is healthy
If there is a bug in the provider the current state of this function wont catch it neither.

> We should first focus on the effort to make dual stack works, imo its good enough for now.
No dout we need to make it work, but this PR is about testing VM's with IPv6

Comment (oshoval) [reply tests/vmi_networking_test.go:889]: i dont agree, bugs can happen in the provider, which will stop making it dual stack, and other stuff wont catch it.
i think we should keep it simple, straight to the point and generic, yet powerful

File: tests/vmi_networking_test.go
Comment (EdDev) [reply tests/vmi_networking_test.go:739]: I think she asked to add "IPv6" in the `By` logs, which you already did. I do not think she asked to mention that twice.

File: tests/vmi_networking_test.go
Comment (EdDev) [inline tests/vmi_networking_test.go:890]: Unfortunately, this will not work because it limits this to running the tests against kubevirtci while we run this against other clusters as well.

This [one](https://github.com/kubevirt/kubevirt/pull/4035/commits/abfb4d07783ee731436e0dc3a2144c4c7fad9357) seem a more  appropriate option although I prefer it through a flag and not an env variable.

For the meantime, I suggest you drop this last commit and instead point to a link on CI that we can see these tests indeed run.
A following PR should add a dedicated test to check dual stack existence, we do not need to check this on every test we run.

**Review by EdDev — CHANGES_REQUESTED** [review]

File: tests/vmi_networking_test.go
Comment (oshoval) [reply tests/vmi_networking_test.go]: dropped

File: tests/vmi_networking_test.go
Comment (oshoval) [reply tests/vmi_networking_test.go]: removed

File: tests/vmi_networking_test.go
Comment (oshoval) [reply tests/vmi_networking_test.go]: removed the comments, if you want please add it later

Comment (ormergi) [reply tests/vmi_networking_test.go]: Why would you do that? we need those comments
It is important to explain to where we ping in IPv6 connectivity test since we dont ping to external web service like in IPv4 connectivity test.

Comment (oshoval) [reply tests/vmi_networking_test.go]: link is broken, if you want me to fix the link please tell what is the live link.

Comment (ormergi) [reply tests/vmi_networking_test.go]: I thought that you added this link at first sorry,
the link is https://github.com/kubevirt/project-infra/blob/master/github/ci/shared-deployments/files/docker-daemon-mirror.conf#L5
it is ok if you dont want to do it in this PR

Comment (oshoval) [reply tests/vmi_networking_test.go]: thx
will add permlink in a commit here

Comment (oshoval) [reply tests/vmi_networking_test.go]: thx
will add permlink in a commit here
hmm but permlink is super long, i will just add the name of the file that have the cidr ?

the permlink
https://github.com/kubevirt/project-infra/blob/a27023bbae7a51a0a35c989c7aa6ede8a40bd6b0/github/ci/shared-deployments/files/docker-daemon-mirror.conf#L5

**Review by EdDev — APPROVED** [review]
Thank you.

File: tests/vmi_networking_test.go
Comment (ormergi) [reply tests/vmi_networking_test.go]: Even though you dropped this commit we will tackle it later so it  is important for me to mention that when you write a test or refactoring one you dont need to worry about the environemnt you need to assume the cluster is functioning with no bugs.
If you want to keep it simple then just check if a pod has both ipv4 and ipv6 addresses, this way you dont bound to any provider.

File: tests/vmi_networking_test.go
Comment (ormergi) [reply tests/vmi_networking_test.go]: Just change the current link with the one I sent, if you dont want to do it now so at least bring back the comments they are important (in this case..)
Dont worry about the lgtm label we will add it again if needed

File: tests/vmi_networking_test.go
Comment (ormergi) [reply tests/vmi_networking_test.go]: You dont realy need to maintain it, it is formatted string with two variable you already have, if you insist then just leave it as is

**Comment by phoracek** [comment]
/approve
/hold

@ormergi remove the hold once you are ok with the PR.

File: tests/vmi_networking_test.go
Comment (ormergi) [inline tests/vmi_networking_test.go:741]: Could you add error messages to `Expect` that call `PingFromVMConsole`?
If wont show the proper error at this state

**Comment by oshoval** [comment]
> @oshoval do we need #3958 first?

Since i tested locally, we dont need it, unless we want to see that prow test it with 2 IPs and then we do need it.
(its robust to auto detect the number of ips)

File: tests/vmi_networking_test.go
Comment (oshoval) [reply tests/vmi_networking_test.go:741]: We will see that PingFromVMConsole failed, as every place that use 
PingFromVMConsole

File: tests/vmi_networking_test.go
Comment (oshoval) [reply tests/vmi_networking_test.go]: thx i prefer to leave it as is.

File: tests/vmi_networking_test.go
Comment (oshoval) [reply tests/vmi_networking_test.go]: done
i am worried because i don't find that important, in order to block PRs and work. 
but updated since its important to you.

**Comment by ormergi** [comment]
/lgtm
/hold
remove hold if you want to merge it now before #3958

**Comment by oshoval** [comment]
/hold cancel
Thanks

**Comment by EdDev** [comment]
> > @oshoval do we need #3958 first?
> 
> Since i tested locally, we dont need it, unless we want to see that prow test it with 2 IPs and then we do need it.
> (its robust to auto detect the number of ips, and the verbosity on prow is limited in case of success anyway).

If we can trust local runs, then we do not need CI at all.
If I understand correctly, the missing PR is required in order to test the IPv6 asserts, therefore if we merge this before that one gets in, we do not test half of the changes done in this PR against the CI.

I see no rush here, please wait until we prove this is working on CI.
/hold

**Comment by ormergi** [comment]
> > > @oshoval do we need #3958 first?
> > 
> > 
> > Since i tested locally, we dont need it, unless we want to see that prow test it with 2 IPs and then we do need it.
> > (its robust to auto detect the number of ips, and the verbosity on prow is limited in case of success anyway).
> 
> If we can trust local runs, then we do not need CI at all.
> If I understand correctly, the missing PR is required in order to test the IPv6 asserts, therefore if we merge this before that one gets in, we do not test half of the changes done in this PR against the CI.
> 
> I see no rush here, please wait until we prove this is working on CI.
> /hold

A PR with lgtm and approve wont merge until all lane are green so you dont need to hold if you want to verify CI will pass

**Comment by EdDev** [comment]
> > I see no rush here, please wait until we prove this is working on CI.
> > /hold
> 
> A PR with lgtm and approve wont merge until all lane are green so you dont need to hold if you want to verify CI will pass

In this case I think it will pass because the dependent PR populates the list of the VMI status IP/s with more than a single IP (the current case). So everything will pass without testing the IPv6 scenarios.

**Comment by oshoval** [comment]
/retest
CI is in bad shape

**Comment by oshoval** [comment]
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4050/pull-kubevirt-e2e-k8s-1.18/1298222567466733568

shows that this PR tests both ips (when rebased on top of #3958)

**Comment by oshoval** [comment]
> https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4050/pull-kubevirt-e2e-k8s-1.18/1298222567466733568
> 
> shows that this PR tests both ips (when rebased on top of #3958)

@EdDev 
can we remove the hold ?
we see in the above link that it works both with 3958, (and shows it use both addresses)
and here that it work without.

**Comment by EdDev** [comment]
> > https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4050/pull-kubevirt-e2e-k8s-1.18/1298222567466733568
> > shows that this PR tests both ips (when rebased on top of #3958)
> 
> @EdDev
> can we remove the hold ?
> we see in the above link that it works both with 3958, (and shows it use both addresses)
> and here that it work without.

This one is dependent on the other one and has no meaning without it.
It may work with your rebase and may not work a minute after when something changed or some conflict got resolved.
I see no rush and you are not suppose to invest time in this anymore.

**Comment by oshoval** [comment]
Rebased since 3958 was merged, removing the hold, PTAL (nothing changed beside automatic rebase)
/hold cancel

**Review by EdDev — APPROVED** [review]

**Comment by oshoval** [comment]
/retest
Thanks @EdDev

**Pull request merged**
